// devexpress-reporting/designer/wizard/internal/_reportWizardStateHelper.js
/**
* DevExpress HTML/JS Reporting (designer\wizard\internal\_reportWizardStateHelper.js)
* Version:  22.2.4
* Build date: Jan 25, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { _convertToStateDataSource } from '../pages/chooseAvailableDataSourcePage';
import { GraphicsUnit } from '../reportWizardState';
export class ReportWizardStateHelper {
    static applyDataBindings(state, model) {
        var modelDS = model.dataSource();
        var dataSourceInfo = modelDS && modelDS['dataSourceInfo'];
        state.dataSource = dataSourceInfo && _convertToStateDataSource(dataSourceInfo);
        state.dataMember = model.dataMember() || '';
    }
    static applyPageSetup(state, model) {
        var unit;
        var _pageSetupProps = [
            model.pageWidth(),
            model.pageHeight(),
            model.margins.left(),
            model.margins.right(),
            model.margins.top(),
            model.margins.bottom()
        ];
        switch (model.measureUnit()) {
            case 'HundredthsOfAnInch':
                _pageSetupProps = _pageSetupProps.map(x => x / 100);
                unit = GraphicsUnit.Inch;
                break;
            case 'TenthsOfAMillimeter':
                _pageSetupProps = _pageSetupProps.map(x => x / 10);
                unit = GraphicsUnit.Millimeter;
                break;
            case 'Pixels':
                unit = GraphicsUnit.Pixel;
                break;
        }
        state.pageSetup = {
            paperKind: model.paperKind(),
            unit: unit,
            width: _pageSetupProps[0],
            height: _pageSetupProps[1],
            marginLeft: _pageSetupProps[2],
            marginRight: _pageSetupProps[3],
            marginTop: _pageSetupProps[4],
            marginBottom: _pageSetupProps[5],
            landscape: model.landscape()
        };
    }
}
